pkg/action/action_lock_store.go

```diff
@@ -0,0 +1,101 @@
+package action
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/turbonomic/kubeturbo/pkg/action/util"
+
+	"github.com/turbonomic/turbo-go-sdk/pkg/proto"
+
+	"github.com/golang/glog"
+	api "k8s.io/client-go/pkg/api/v1"
+)
+
+type IActionLockStore interface {
+	getLock(actionItem *proto.ActionItemDTO) (*util.LockHelper, error)
+}
+
+type ActionLockStore struct {
+	// The lock map for concurrent control of action execution
+	lockMap *util.ExpirationMap
+
+	// The function to get the related pod from action item
+	podFunc func(ai *proto.ActionItemDTO) *api.Pod
+}
+
+func newActionLockStore(lockMap *util.ExpirationMap, podFunc func(ai *proto.ActionItemDTO) *api.Pod) *ActionLockStore {
+	return &ActionLockStore{lockMap, podFunc}
+}
+
+const (
+	defaultWaitLockTimeOut = time.Second * 300
+	defaultWaitLockSleep   = time.Second * 10
+)
+
+// Acquires the lock for the action item. It will wait and retry if the lock is not available, i.e.,
+// the lock is used by other action item.
+// The key used to acquire the lock is as follows:
+//
+// 1. If the action item is associated to a container pod (meaning, the function podFunc returns a pod),
+//    the key is the pod id for bare-pod cases and its parent controller id for non-bare-pod cases.
+//
+// 2. Otherwise, the key is the id of the target SE of the action item.
+func (a *ActionLockStore) getLock(actionItem *proto.ActionItemDTO) (*util.LockHelper, error) {
+	id := actionItem.GetUuid()
+	if key, err := a.getLockKey(actionItem); err != nil {
```

Comment:
I think it can not handle bare-Pod case well:
    Two actions on the same bare-Pod,  then the key for them will be same. The first action will update the cache, while the second action will use the same key to lock the bare-Pod.

Problem is that:  During the execution of second action,  a third action comes on the same bare-Pod, then the third action will use another key to get a lock, and it will get the lock. Then the second and third actions will both working on the same Pod.

Reply:
For bare-Pod case, we should try to get the lock with the latest pod name